### DISCUSSION TITLE
Multi vehicle topic pattern matching message path syntax extension

### DISCUSSION THREAD

**Discussion by mhl787156** [foxglove/community#210] [Q&A]
Hi! 

So I'm trying to use foxglove with my ROS2 multi-vehicle use case where each vehicle has its own namespaced set of topics (e.g. `/vehicle_1/mavros/local_position/pose`, `/vehicle_2/mavros/local_position/pose`). One of the interesting aspects is that I can have a variable number of vehicles coming and going from the network at any time. 

It would be cool to have foxglove be able to present position or other data in a dynamic manner. Currently you need to write the complete topic name, and change all references to them every time we wish to monitor a different vehicle. Perhaps there is scope for pattern matched topics? e.g. `/vehicle_*/mavros/local_position/pose.position.x` which would start plotting positions of any vehicles that come by! 

I see that current message path syntax supports filtering by an id, but that would require a substantial redesign of our current setup.

I just thought i'd float the idea across - we've been using foxglove somewhat successfully for single drone cases otherwise! 

Many Thanks

**Comment by amacneil** [comment]
FYI you can already do this in some panels (including plots) using [variables](https://foxglove.dev/docs/studio/app-concepts/variables). However, it's not supported in the 3d panel or image panels.

Let us know if that helps!

**Comment by jtbandes** [comment]
I don't think the variables are quite like what is being requested, because you can't have the topic name itself be dynamically chosen based on a variable (only slices and filters), and a variable can only have one value at a time (unlike the `*` wildcard in the example).

**Comment by mhl787156** [comment]
@amacneil Thanks for you reply, but I agree with @jtbandes, variables will allow me to specify a specific vehicle_id and perhaps change that on the fly. But that will require me to ensure that messages themselves have the `id` field to filter by. 

However as jtbandes says, the topic name itself cannot be dynamic, which is what I need for my usecase. From discussions with my colleagues, its a 50:50 split on how multi-vehicle ROS should be architected between namespacing topics, and id-ing messages.

**Comment by amacneil** [comment]
I see, I missed the part where you wanted _all_ vehicles to show up.

Supporting glob patterns seems like a nice idea. It would be much easier to fit into the UX of our raw message panels and other panels where you type message path directly, but fitting that into our 3d panel (which uses a topic selector) will require some re-thinking.

cc @2metres
